hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/keyvalue/TestContainerReconciliationWithMockDatanodes.java

```diff
@@ -305,6 +310,30 @@ public void testContainerReconciliationWithPeerFailure(FailureLocation failureLo
     mockContainerProtocolCalls();
   }
 
+  @Test
+  public void testContainerReconciliationFailureContainerScan()
+      throws Exception {
+    // Use synchronous on-demand scans to re-build the merkle trees after corruption.
+    datanodes.forEach(d -> d.scanContainer(CONTAINER_ID));
+
+    // Each datanode should have had one on-demand scan during test setup, and a second one after corruption was
+    // introduced.
+    waitForExpectedScanCount(1);
+
+    for (MockDatanode current : datanodes) {
+      doThrow(IOException.class).when(current.getHandler().getChecksumManager()).read(any());
+      List<DatanodeDetails> peers = datanodes.stream()
+          .map(MockDatanode::getDnDetails)
+          .filter(other -> !current.getDnDetails().equals(other))
+          .collect(Collectors.toList());
+      // Reconciliation should fail for each datanode, since the checksum info cannot be retrieved.
+      assertThrows(IOException.class, () -> current.reconcileContainer(dnClient, peers, CONTAINER_ID));
+      Mockito.reset(current.getHandler().getChecksumManager());
+    }
+    // Even failure of Reconciliation should have triggered a second on-demand scan for each replica.
+    waitForExpectedScanCount(2);
+  }
```

Comment:
This version of the test also passes, but does not require extra changes to accommodate the spy.
```suggestion
  @Test
  public void testContainerReconciliationFailureContainerScan() throws Exception {
    for (MockDatanode current : datanodes) {
      List<DatanodeDetails> peers = datanodes.stream()
          .map(MockDatanode::getDnDetails)
          .filter(other -> !current.getDnDetails().equals(other))
          .collect(Collectors.toList());
      FileUtils.deleteDirectory(new File(current.getContainer(CONTAINER_ID).getContainerData().getMetadataPath()));
      // Reconciliation should fail without a metadata directory present.
      assertThrows(IOException.class, () -> current.reconcileContainer(dnClient, peers, CONTAINER_ID));
      Mockito.reset(current.getHandler().getChecksumManager());
    }
    // Each replica should have an on-demand scan triggered when reconciliation failed.
    waitForExpectedScanCount(1);
    datanodes.forEach(d -> assertTrue(d.getContainer(CONTAINER_ID).getContainerData().isUnhealthy()));
  }
```

Reply:
This passes in isolation for me locally, but I forgot this test re-uses container instances across tests to make sure the data is truly repaired. Mocking the fault injection like the current code does is probably the easier approach here.